src/lib/merkle_map.ts

```diff
@@ -0,0 +1,126 @@
+import {
+  Experimental,
+  isReady,
+  shutdown,
+  Field,
+  Bool,
+  SelfProof,
+  CircuitValue,
+  arrayProp,
+  Poseidon,
+  prop,
+  Circuit,
+} from 'snarkyjs'
+
+import assert from 'assert';
+
+const bits = 255;
+
+const printDebugs = false;
+
+export class MerkleMap {
+  tree: InstanceType<typeof Experimental.MerkleTree>;
+
+  // ------------------------------------------------
+
+  constructor() {
+    assert(bits <= 255, 'bits must be <= 255')
+    if (bits != 255) {
+      console.warn('bits set to', bits + '. Should be set to 255 in production to avoid collisions');
+    }
+    this.tree = new Experimental.MerkleTree(bits+1);
+  }
+
+  // ------------------------------------------------
+
+  _keyToIndex(key: Field) {
+    // the bit map is reversed to make reconstructing the key during proving more convenient
+    let keyBits = key.toBits().slice(0,bits).reverse().map((b) => b.toBoolean());
+
+    var n = BigInt(0);
+    for (var i = 0; i < keyBits.length; i++) {
+      var b = keyBits[i] ? 1 : 0;
+      n += BigInt(2)**BigInt(i) * BigInt(b);
+    }
+
+    return BigInt(n)
+  }
+
+  // ------------------------------------------------
+
+  set(key: Field, value: Field) {
+    const index = this._keyToIndex(key);
+    this.tree.setLeaf(index, value);
+  }
+
+  // ------------------------------------------------
+
+  get(key: Field) {
+    const index = this._keyToIndex(key);
+    return this.tree.getNode(0, index);
+  }
+
+  // ------------------------------------------------
+
+  getRoot() {
+    return this.tree.getRoot();
+  }
+
+  getWitness(key: Field) {
+    const index = this._keyToIndex(key);
+    const value = this.tree.getNode(0, index);
+    const root = this.tree.getRoot();
+
+    class MerkleWitness extends Experimental.MerkleWitness(bits + 1) {}
+
+    const witness = new MerkleWitness(this.tree.getWitness(index));
+
+    if (printDebugs) {
+      // witness bits and key bits should be the reverse of each other, so
+      // we can calculate the key during recursively traversing the path
+      console.log('witness bits', witness.isLeft.map((l) => l.toBoolean() ? '0' : '1').join(', '));
+      console.log('key bits', key.toBits().slice(0,bits).map((l) => l.toBoolean() ? '1' : '0').join(', '));
```

Comment:
this confuses me. in `_keyToIndex`, I thought the bit sequence was reversed, i.e. `11001` -> `10011`
Here, instead it seems like a bitwise NOT is applied: `11001` -> `00110`, giving a different result, or am I missing something?

Reply:
I think both are true actually - since both `isLeft` is reading out `true -> 0` and `false -> 1` - and computing the path from `isLeft` is computing the key in the "opposite" order than the key itself was entered (so the reversing step in `_keyToIndex`)

Reply:
Alright, thanks!